packages/shared_preferences/shared_preferences_android/android/src/main/kotlin/io/flutter/plugins/sharedpreferences/SharedPreferencesPlugin.kt

```diff
@@ -0,0 +1,306 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.plugins.sharedpreferences
+
+import android.content.Context
+import android.util.Base64
+import android.util.Log
+import androidx.annotation.VisibleForTesting
+import androidx.datastore.core.DataStore
+import androidx.datastore.preferences.core.Preferences
+import androidx.datastore.preferences.core.booleanPreferencesKey
+import androidx.datastore.preferences.core.doublePreferencesKey
+import androidx.datastore.preferences.core.edit
+import androidx.datastore.preferences.core.longPreferencesKey
+import androidx.datastore.preferences.core.stringPreferencesKey
+import androidx.datastore.preferences.preferencesDataStore
+import io.flutter.embedding.engine.plugins.FlutterPlugin
+import io.flutter.plugin.common.BinaryMessenger
+import java.io.ByteArrayInputStream
+import java.io.ByteArrayOutputStream
+import java.io.ObjectInputStream
+import java.io.ObjectOutputStream
+import java.lang.Exception
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.firstOrNull
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.runBlocking
+
+const val TAG = "SharedPreferencesPlugin"
+const val SHARED_PREFERENCES_NAME = "FlutterSharedPreferences"
+const val LIST_PREFIX = "VGhpcyBpcyB0aGUgcHJlZml4IGZvciBhIGxpc3Qu"
+const val DOUBLE_PREFIX = "VGhpcyBpcyB0aGUgcHJlZml4IGZvciBEb3VibGUu"
+
+private val Context.sharedPreferencesDataStore: DataStore<Preferences> by
+    preferencesDataStore(SHARED_PREFERENCES_NAME)
+
+/// SharedPreferencesPlugin
+class SharedPreferencesPlugin() : FlutterPlugin, SharedPreferencesAsyncApi {
+  private lateinit var context: Context
+
+  private var listEncoder = ListEncoder() as SharedPreferencesListEncoder
+
+  @VisibleForTesting
+  constructor(listEncoder: SharedPreferencesListEncoder) : this() {
+    this.listEncoder = listEncoder
+  }
+
+  private fun setUp(messenger: BinaryMessenger, context: Context) {
+    this.context = context
+    try {
+      SharedPreferencesAsyncApi.setUp(messenger, this)
+    } catch (ex: Exception) {
+      Log.e(TAG, "Received exception while setting up SharedPreferencesPlugin", ex)
+    }
+  }
+
+  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
+    setUp(binding.binaryMessenger, binding.applicationContext)
+    DeprecatedSharedPreferencesPlugin().onAttachedToEngine(binding)
+  }
+
+  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
+    SharedPreferencesAsyncApi.setUp(binding.binaryMessenger, null)
+  }
+
+  /** Adds property to data store of type bool. */
+  override fun setBool(
+      key: String,
+      value: Boolean,
+      options: SharedPreferencesPigeonOptions
+  ): Boolean {
+    return runBlocking { dataStoreSetBool(key, value) }
+  }
+
+  private suspend fun dataStoreSetBool(key: String, value: Boolean): Boolean {
```

Comment:
n00b Kotlin question: we can't just inline these into the `runBlocking` lambdas?